runner/src/runner.rs

```diff
@@ -0,0 +1,631 @@
+    fn compute_storage_key(&self, module: String, key: String) -> String {
+        let module = twox_128(module.as_bytes());
+        let item = twox_128(key.as_bytes());
+        let key = hex::encode([module, item].concat());
+        format!("0x{}", key)
+    }
+
+    async fn query_storage(
+        &self,
+        parachain_rpc: &WebsocketClient,
+        maybe_storage_key: Option<String>,
+        method: String,
+    ) -> Option<SpCoreBytes> {
+        let params = maybe_storage_key.map_or(rpc_params![], |key| rpc_params![key]);
+        parachain_rpc.inner().request(method.as_str(), params).await.ok()
+    }
+
+    async fn read_chain_storage<T: 'static + Decode + Debug>(
+        &self,
+        parachain_rpc: &WebsocketClient,
+        maybe_storage_key: Option<String>,
+    ) -> Result<Option<T>, Error> {
+        read_chain_storage(self, parachain_rpc, maybe_storage_key).await
+    }
+}
+
+pub async fn ws_client(url: &str) -> Result<WebsocketClient, Error> {
+    let ws_client = WsClientBuilder::default().build(url).await?;
+    Ok(WebsocketClient::new(ws_client))
+}
+
+async fn download_binary(runner: &impl RunnerExt, release: ClientRelease) -> Result<DownloadedRelease, Error> {
+    let (bin_name, bin_path) = runner.uri_to_bin_path(&release.uri)?;
+    log::info!("Downloading {} at: {:?}", bin_name, bin_path);
+    File::create(bin_path.clone())?;
+
+    let bytes = runner.get_request_bytes(release.uri.clone()).await?;
+    fs::write(&bin_path, &bytes)?;
+
+    // Make the binary executable.
+    // The set permissions are: -rwx------
+    fs::set_permissions(bin_path.clone(), fs::Permissions::from_mode(0o700))?;
+
+    Ok(DownloadedRelease {
+        release,
+        path: bin_path,
+        bin_name: bin_name.to_string(),
+    })
+}
+
+fn uri_to_bin_path(runner: &impl RunnerExt, uri: &str) -> Result<(String, PathBuf), Error> {
+    // Remove any trailing slashes from the release URI
+    let parsed_uri = Url::parse(uri.trim_end_matches('/'))?;
+    let bin_name = parsed_uri
+        .path_segments()
+        .and_then(|segments| segments.last())
+        .and_then(|name| if name.is_empty() { None } else { Some(name) })
+        .ok_or(Error::ClientNameDerivationError)?;
+    let bin_path = runner.download_path().join(bin_name);
+    Ok((bin_name.to_string(), bin_path))
+}
+
+fn delete_downloaded_release(runner: &impl RunnerExt) -> Result<(), Error> {
+    let release = runner.downloaded_release().as_ref().ok_or(Error::NoDownloadedRelease)?;
+    log::info!("Removing old release, with path {:?}", release.path);
+    fs::remove_file(&release.path)?;
+    Ok(())
+}
+
+fn terminate_proc_and_wait(runner: &mut impl RunnerExt) -> Result<u32, Error> {
+    let child_proc = runner.child_proc().as_mut().ok_or(Error::NoChildProcess)?;
+    signal::kill(
+        Pid::from_raw(child_proc.id().try_into().map_err(|_| Error::IntegerConversionError)?),
+        Signal::SIGTERM,
+    )?;
+
+    match child_proc.wait() {
```

Comment:
You know how some processes refuse to close after receiving a kill signal? Do we have to worry about / deal with that?

Reply:
According to the docs, the `wait` call synchronously waits for the process to finish. The one [warning](https://doc.rust-lang.org/std/process/struct.Child.html#warning) in the `Child` struct introduction is specifically about the requirement to `wait()` for OS resources to be free.

In the worst case, the runner will hang for a while on this line, waiting for the vault binary to end. I think this is correct behaviour (sending `SIGTERM` rather than `SIGKILL`), since we want the Bitcoin node connection to be terminated gracefully to avoid potential double-spends.

Reply:
What happens if there is a Vault already running when the runner starts?

Reply:
This function only kills the child process spawned by the runner (`runner.child_proc()`). Since the runner has no knowledge about the PIDs of other running vaults, it can't kill them. If an existing vault process was spawned from an executable that's in the `download_path` and has the same name, the executable would be overwritten if it's not the latest (but the process spawned from the executable would be unaffected).

Reply:
Does the vault stop when you stop the `vault-runner`? E.g. when you ctrl+c the process

Reply:
I had thought that killing a parent process kills children as well but this doesn't seem to be the case. This is an edge that should be handled, I'll try to come up with a unit test for it.